clients/pkg/promtail/promtail.go

```diff
@@ -162,7 +166,20 @@ func (p *Promtail) reloadConfig(cfg *config.Config) error {
 		}
 	}
 
-	tms, err := targets.NewTargetManagers(p, p.reg, p.logger, cfg.PositionsConfig, p.client, cfg.ScrapeConfig, &cfg.TargetConfig)
+	// this is the sink were all scraped log entries should get  to
+	var entriesSink api.EntryHandler = p.client
+
+	// If WAL is enabled, instantiate the WAL itself, it's writer, and use that as entries sink for all scraping targets
+	if cfg.WAL.Enabled {
+		pWAL, err := wal.New(cfg.WAL, p.logger, p.reg)
+		if err != nil {
+			return fmt.Errorf("error starting WAL: %w", err)
+		}
+		p.walWriter = wal.NewWriter(pWAL, p.logger, p.client)
+		entriesSink = p.walWriter
+	}
+
+	tms, err := targets.NewTargetManagers(p, p.reg, p.logger, cfg.PositionsConfig, entriesSink, cfg.ScrapeConfig, &cfg.TargetConfig)
```

Comment:
I think it probably makes more sense to have promtail have a field that is an entry handler, and during the loading of the config here we create an entry handler that wraps both the list of clients and the walWriter, rather than doing the propagation of the channel data within the wal writer.

Reply:
Yhea same, wasn't that happy with how it looked adding the fanout inside the writer. Moved it to promtail.go, and added a new slim component that does the fanout.

Reply:
In [this commit](https://github.com/grafana/loki/pull/8267/commits/5af6e20ec688caf4e327f32346680f1626c02061) the fanout is moved out of the `Writer`, and done in `promtail.go` via a `EntryHandlerFanouter`, component that takes n entry handlers and exposes a single one, performing a fanout. Could be re-used when the `client.Manager` needs to implement the same functionality as the current multiclient.